# Day 3 In-class Assignment: The power of compound interest
<img src="https://g.foolcdn.com/image/?url=https%3A//g.foolcdn.com/editorial/images/463624/retirement-savings-jar-full-of-coins-and-alarm-clock.jpg&w=2000&op=resize" width=400px>


### <p style="text-align: right;"> &#9989; Put your name here.
<p style="text-align: right;"> &#9989; Put your group member names here.

## Learning Goals:

By the end of this assignment you should be able to:

* Use Python lists to store values
* Use loops to iterate over a specified range or until a condition is met.
* Compute the best retirement savings plan.


---
## The Problem

You are working for a bank, writing code to calculate the revenue and expenses for the company. A person can deposit money in a retirement account at your bank, and the bank pays interest every year on the money in the retirement account. The mathematical expression for calculating the amount of money in the retirement account every year is:

$\mathrm{New~Retirement~Total} = (\mathrm{Old~Retirement~Total}) + [(\mathrm{Old~Retirement~Total})*(\mathrm{Interest~Rate})] + (\mathrm{Money~Added~to~Retirement})$

Or, if we simplify this a bit, we can re-write it as:

$\mathrm{New~Retirement~Total} = [(\mathrm{Old~Retirement~Total})*(\mathrm{1 + Interest~Rate})] + (\mathrm{Money~Added~to~Retirement})$

**Your job is to write a piece of code that calculates and stores (in a list) the total amount of money in the retirement account *every year* for some number of years** (your boss is going to make a plot of it, so make sure you have the value for each year!). 

Your boss tells you to model the total retirement savings for a ***range*** of different parameters (*e.g.* interest rates and annual contributions).

**You can assume that the person saving for retirement starts when they're 18 and continues until they're 65. Also assume that the interest rate and annual contributions are fixed.**

---
## 1. Planning your solution and pseudo-coding

Before you start writing any code you should come up with a plan for how you've going to solve the problem. During this phase you should identify and define the variables that are revelant to the problem, the equations you're going to need to model, the overall structure of the program that you will write, and how you will make sure the code is running correctly.

#### You and your group members are expected to use the white board at your table (or a virtual whiteboard app like [Google Jamboard](https://jamboard.google.com/) or the [Aww app](https://awwapp.com/)) to write out your ideas to ensure that everyone is on the same page and allow your instructors to see your progress.

Part of this planning process should involve writing "**pseudo-code**", a plain language description of the steps in your code (sort of like writing out in your code in form of a cooking recipe). **You actually already did a version of this on Day 1 when you and your group came up with an _algorithm_ for sorting people by their birthday.**

As a reminder, **an algorithm is**: a procedure for solving a problem in terms of the actions to be executed and the order in which those actions are to be executed. In other words, an algorithm is merely the sequence of steps taken to solve a problem. [[source](https://www.unf.edu/~broggio/cop2221/2221pseu.htm) of this definition]

You can see some examples of pseudocode [here](https://www.unf.edu/~broggio/cop2221/2221pseu.htm) and [here](https://www.geeksforgeeks.org/how-to-write-a-pseudo-code/).

Things you should decide on during your planning and psuedo-coding:

1. What quantities in your model are "variables" that will change as a function of time? What are the initial values of these variables?
2. What quantities in your model are "parameters" that are constant over time? What are reasonable ranges for each parameter? 
4. How will you go about solving this problem for a given set of inputs?
5. How will you check to make sure your code is running correctly. 


You should be able to justify your choices.

### &#9989;&nbsp; 1.1 Record the details of you plan in the Markdown cell below.

<font size=6 color="#009600">&#9998;</font> *year, retTot are moving variables, and intRate and newMoney.  Range from 18-65, which is 47 years (indicies) 
retTot = retTot*(1+IntRate)+newMoney*

## &#128721; STOP AND ASK YOUR INSTRUCTOR TO COME OVER AND CHECK YOUR PSEUDOCODE

---
## 2. Coding up your solution

Now that you and your group have come up with a plan for your code, it's time to put your plan into action.

### &#9989;&nbsp; 2.1 Writing Code
**Write your code for calculating the total retirement savings in the cell below. Start by writing code for a single set of parameter values (i.e., not a range).** It is simpler, and you can also add to it later.

**Hint: Remember to update the value of $\mathrm{Old~Retirement~Total}$ in your loop!**

In [17]:
# Put your code for calculating the total amount of retirement savings here
retTotList=[]
retTot=0
interestRate = .05
newMoney = 1000
for year in range(48):
    print(year+18,retTot)
    retTot= (retTot*(1+interestRate))+newMoney
    retTotList.append(retTot)
    
print(retTotList)
    
    


18 0
19 1000.0
20 2050.0
21 3152.5
22 4310.125
23 5525.63125
24 6801.912812500001
25 8142.008453125001
26 9549.108875781252
27 11026.564319570316
28 12577.892535548832
29 14206.787162326274
30 15917.126520442589
31 17712.982846464718
32 19598.631988787954
33 21578.563588227353
34 23657.491767638723
35 25840.36635602066
36 28132.384673821693
37 30539.00390751278
38 33065.95410288842
39 35719.251808032845
40 38505.21439843449
41 41430.47511835621
42 44501.998874274024
43 47727.098817987724
44 51113.45375888711
45 54669.12644683147
46 58402.582769173045
47 62322.7119076317
48 66438.8475030133
49 70760.78987816397
50 75298.82937207217
51 80063.77084067579
52 85066.95938270958
53 90320.30735184507
54 95836.32271943733
55 101628.1388554092
56 107709.54579817966
57 114095.02308808865
58 120799.77424249309
59 127839.76295461775
60 135231.75110234864
61 142993.33865746608
62 151143.0055903394
63 159700.15586985636
64 168685.16366334917
65 178119.42184651663
[1000.0, 2050.0, 3152.5, 4310.125, 55

---
### &#128721; STOP

**Check to make sure your code is working correctly.**

In [ ]:
# If your method for checking your results involves writing code, put it here

*Explain how you know your code is working correctly.*

<font size=6 color="#009600">&#9998;</font> *Put your explanation here!* 
calculator!
1000*1.05+1000=2050
2050*1.05+1000=3152.5
yay

## 3. Coding up your solution (Range of Values)

### &#9989;&nbsp; 3.1 Putting in a Range of Values
**Take the code you wrote in Part 2 and expand on it so that you can calculate total retirement savings for a range of different input parameters.** To start, try 3 different annual contribution values and 4 different interests rates, but try to write your code so that you could easily do greater or fewer numbers of values for both parameters.

**Note:** To explore a range of input parameters, you may want to use a "nested" loop (a loop inside a loop). Look back at the pre-class assignment for an example!

In [46]:
startAge=18
endAge=65
finalRetTotalList = []
intRateList = [0.03, 0.05, 0.07]
newMoneyList = [1000, 2000, 3000]
for newMoney in newMoneyList:
    varNewMoney = []
    retTotal = 15000
    for intRate in intRateList:
        varRateList = []
        retTotalCopy = retTotal
        for i in range(endAge+1-startAge):
            retTotalCopy = (retTotalCopy*(1+intRate)) + newMoney
            varRateList.append(int(retTotalCopy))
        varNewMoney.append(varRateList)
    finalRetTotalList.append(varNewMoney)
for i in finalRetTotalList:
    print(i)

[[16450, 17943, 19481, 21066, 22698, 24379, 26110, 27893, 29730, 31622, 33571, 35578, 37645, 39775, 41968, 44227, 46554, 48950, 51419, 53962, 56580, 59278, 62056, 64918, 67865, 70901, 74028, 77249, 80567, 83984, 87503, 91128, 94862, 98708, 102670, 106750, 110952, 115281, 119739, 124331, 129061, 133933, 138951, 144120, 149443, 154927, 160574, 166392], [16750, 18587, 20516, 22542, 24669, 26903, 29248, 31710, 34296, 37011, 39861, 42854, 45997, 49297, 52762, 56400, 60220, 64231, 68443, 72865, 77508, 82384, 87503, 92878, 98522, 104448, 110670, 117204, 124064, 131267, 138831, 146772, 155111, 163867, 173060, 182713, 192849, 203491, 214666, 226399, 238719, 251655, 265238, 279500, 294475, 310199, 326708, 344044], [17050, 19243, 21590, 24101, 26789, 29664, 32740, 36032, 39554, 43323, 47356, 51671, 56288, 61228, 66514, 72170, 78222, 84698, 91626, 99040, 106973, 115461, 124544, 134262, 144660, 155786, 167691, 180430, 194060, 208644, 224249, 240947, 258813, 277930, 298385, 320272, 343691, 368750, 3

## 4. Bringing in new customers

Your boss says the bank is trying to attract customers that don't start saving for retirement until later in life, and they're trying to find a good way to do this. You suggest offering **double the usual interest rate** for people that don't start saving until they're 30. 

Your boss likes the idea and wants you to run the numbers. 

### &#9989;&nbsp; 4.1 Answering your Boss' Question, Code
**Who would end up with more money saved: someone that starts saving at 18 with a normal interest rate, or someone that starts saving at 30 with double the interest rate?**

In [53]:
# Put your code here
endAge=65
startAgeList=[18,30]
finalRetTotalList = []
intRateList = [0.03, 0.06]
newMoneyList = [1000]
for newMoney in newMoneyList:
    varNewMoney = []
    retTotal = 15000
    for intRate in intRateList:
        varRateList = []
        retTotalCopy = retTotal
        for startAge in startAgeList:
            varStartList = []
            retTotalCopy = retTotal
        for i in range(endAge+1-startAge):
            retTotalCopy = (retTotalCopy*(1+intRate)) + newMoney
            varRateList.append(int(retTotalCopy))
        varNewMoney.append(varRateList)
    finalRetTotalList.append(varNewMoney)
for i in finalRetTotalList:
    print(i)

    #241329 > 106750

[[16450, 17943, 19481, 21066, 22698, 24379, 26110, 27893, 29730, 31622, 33571, 35578, 37645, 39775, 41968, 44227, 46554, 48950, 51419, 53962, 56580, 59278, 62056, 64918, 67865, 70901, 74028, 77249, 80567, 83984, 87503, 91128, 94862, 98708, 102670, 106750], [16900, 18914, 21048, 23311, 25710, 28253, 30948, 33805, 36833, 40043, 43446, 47052, 50876, 54928, 59224, 63777, 68604, 73720, 79143, 84892, 90986, 97445, 104292, 111549, 119242, 127397, 136040, 145203, 154915, 165210, 176123, 187690, 199952, 212949, 226726, 241329]]


### &#9989;&nbsp; 4.2 Answering your Boss' Question, Answer

**Who would end up saving more money for retirement?**

*The 18 year old will make more money, as the original value would have compounded much more than the 30 year old's over time, even if the rates were doubled.*

---
## Assignment wrap-up

Please fill out the form that appears when you run the code below. ***For today's assignment, ignore the part about pre-class assignment.*  You must completely fill this out in order to receive credit for the assignment!** 

In [54]:
from IPython.display import HTML
HTML(
"""
<iframe 
	src="https://cmse.msu.edu/cmse201-ic-survey" 
	width="800px" 
	height="600px" 
	frameborder="0" 
	marginheight="0" 
	marginwidth="0">
	Loading...
</iframe>
"""
)

## Congratulations, you're done!

Submit this assignment by uploading it to the course Desire2Learn web page.  Go to the "In-class assignments" folder, find the appropriate submission link, and upload it there.

See you next class!

&#169; Copyright 2023,  Department of Computational Mathematics, Science and Engineering at Michigan State University.